mlkem/poly.h

```diff
@@ -215,8 +215,27 @@ ENSURES(ARRAY_IN_BOUNDS(int, k, 0, (MLKEM_N - 1), r->coeffs, 0, 4095));
 
 #define poly_frommsg MLKEM_NAMESPACE(poly_frommsg)
 void poly_frommsg(poly *r, const uint8_t msg[MLKEM_INDCPA_MSGBYTES]);
+
+
 #define poly_tomsg MLKEM_NAMESPACE(poly_tomsg)
-void poly_tomsg(uint8_t msg[MLKEM_INDCPA_MSGBYTES], const poly *r);
+/*************************************************
+ * Name:        poly_tomsg
+ *
+ * Description: Convert polynomial to 32-byte message
+ *
+ * Arguments:   - uint8_t *msg: pointer to output message
+ *              - const poly *r: pointer to input polynomial
+ *                Coefficients must be unsigned canonical
+ **************************************************/
+void poly_tomsg(uint8_t msg[MLKEM_INDCPA_MSGBYTES], const poly *r)
+    // clang-format off
+REQUIRES(msg != NULL && IS_FRESH(msg, MLKEM_INDCPA_MSGBYTES))
+REQUIRES(r != NULL && IS_FRESH(r, sizeof(poly)))
```

Comment:
@rod-chapman Could you attempt to show for this one that it's functionally indeed equivalent to
```c
          // t += ((int16_t)t >> 15) & MLKEM_Q
          // t  = (((t << 1) + MLKEM_Q/2)/MLKEM_Q) & 1;
```

You may want/need to hoist the loop body into a separate function operating on scalars for that.

Reply:
I changed the body of the loop to:
```
          uint32_t coeff = a->coeffs[8 * i + j];
          // t += ((int16_t)t >> 15) & MLKEM_Q;
          // t  = (((t << 1) + MLKEM_Q/2)/MLKEM_Q) & 1;
          uint32_t t = coeff << 1;
          t += 1665;
          t *= 80635;
          t >>= 28;
          t &= 1;

          ASSERT(t == ((((coeff + (((int16_t)coeff >> 15) & MLKEM_Q)) << 1) +
                        MLKEM_Q / 2) /
                       MLKEM_Q) &
                     1,
                 "axiomatic definition of Compress1");
```
but both Z3 and Bitwuzla fails to prove that. I can prove that these two forms are equivalent by exhaustive test over values 0 .. 3328, though.

Reply:
Note that the sequence of assignments to t can actually be replaced by:
```
t = (coeff >= 833) && (coeff <= 2496);
```
would that be a justifiable simplification (and performance improvement)?

Reply:
But I have concern that a compiler like LLVM might not respect constant time on that expression and its subsequent use to assign to `msg[i]` ??

Reply:
Belay that... I confused the precedence of == and &. Added parentheses makes it prove OK.

Reply:
Done and pushed.